File: src/algorithms/cpp/Super_Ugly_Number.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
#include <queue>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
public:
    int nthSuperUglyNumber(int n, vector<int>& primes) {
        vector <long> uglys (1, 1), ugly_indexes (SZ(primes), 0);
        priority_queue <pair<long, long>, vector <pair<long, long>>, greater<pair<long, long>>> heap;
        for (int i = 0; i < SZ(primes); ++ i)
            heap.push(make_pair(primes[i], i));
        for (int i = 0; i < n; ++ i) {
            pair <long, long> minn = heap.top();
            heap.pop();
            if (minn.first != uglys[SZ(uglys) - 1])   uglys.push_back(minn.first);
            else                                      -- i;
            heap.push(make_pair(uglys[++ ugly_indexes[minn.second]] * primes[minn.second], minn.second));
        }
        return uglys[n - 1];
    }
};

int main() {
    return 0;
}

File: src/algorithms/python/Super_Ugly_Number.py
# -*- coding: utf-8 -*-


class Solution(object):
    def nthSuperUglyNumber(self, n, primes):
        """
        :type n: int
        :type primes: List[int]
        :rtype: int
        """
        ugly_nums, prime_indexs = [1], [0 for i in range(len(primes))]
        for i in range(n - 1):
            minn = primes[0] * ugly_nums[prime_indexs[0]]
            for j in range(1, len(primes)):
                minn = min(minn, primes[j] * ugly_nums[prime_indexs[j]])
            for j in range(0, len(primes)):
                if primes[j] * ugly_nums[prime_indexs[j]] == minn:
                    prime_indexs[j] += 1
            ugly_nums.append(minn)
        return ugly_nums[-1]
